src/IBM.WatsonDeveloperCloud/Sockets/AWatsonWSClient.cs

```diff
@@ -0,0 +1,153 @@
+﻿/**
+* Copyright 2017 IBM Corp. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+*/
+
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.WebSockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace IBM.WatsonDeveloperCloud.Sockets
+{
+    public abstract class AWatsonWSClient
+    {
+        ArraySegment<byte> closingMessage = new ArraySegment<byte>(Encoding.UTF8.GetBytes(
+           "{\"action\": \"stop\"}"
+       ));
+
+        private const int ReceiveChunkSize = 1024;
+        private const int SendChunkSize = 1024;
+
+        protected ClientWebSocket BaseClient { get; set; }
+        protected UriBuilder UriBuilder { get; set; }
+        protected Dictionary<string, string> QueryString { get; set; }
+
+        public Action OnOpen = () => { };
+        public Action<string> OnMessage = (message) => { };
+        public Action<Exception> OnError = (ex) => { };
+        public Action OnClose = () => { };
+
+        public AWatsonWSClient AddArgument(string argumentName, string argumentValue)
+        {
+            if (QueryString.ContainsKey(argumentName))
+                QueryString[argumentName] = argumentValue;
+            else
+                QueryString.Add(argumentName, argumentValue);
+
+            UriBuilder.Query =
+                string.Join("&", QueryString.Keys.Where(key => !string.IsNullOrWhiteSpace(QueryString[key])).Select(key => string.Format("{0}={1}", key, QueryString[key])));
```

Comment:
This is probably something we should encode because strings with spaces will break the url